api/src/services/generate-xform.js

```diff
@@ -0,0 +1,162 @@
+    let stdout = '';
+    let stderr = '';
+    xsltproc.stdout.on('data', data => stdout += data);
+    xsltproc.stderr.on('data', data => stderr += data);
+    xsltproc.stdin.setEncoding('utf-8');
+    xsltproc.stdin.write(formXml);
+    xsltproc.stdin.end();
+    xsltproc.on('close', code => {
+      if (code !== 0 || stderr.length) {
+        logger.error('xsltproc stderr output: ');
+        logger.error(stderr);
+        return reject(new Error(`Error transforming xml, xsltproc returned code "${code}"`));
+      }
+      resolve(stdout);
+    });
+  });
+};
+
+const generateForm = formXml => {
+  return transform(formXml, FORM_STYLESHEET).then(form => {
+    return form.replace(FORM_ROOT_OPEN, '')
+               .replace(ROOT_CLOSE, '')
+               .replace(JAVAROSA_SRC, ' data-media-src="');
+  });
+};
+
+const generateModel = formXml => {
+  return transform(formXml, MODEL_STYLESHEET).then(model => {
+    return model.replace(MODEL_ROOT_OPEN, '')
+                .replace(ROOT_CLOSE, '');
+  });
+};
+
+const getEnketoForm = doc => {
+  const collect = doc.context && doc.context.collect;
+  return !collect && formsService.getXFormAttachment(doc);
+};
+
+const generate = formXml => {
+  return Promise.all([ generateForm(formXml), generateModel(formXml) ])
+    .then(([ form, model ]) => ({ form, model }));
+};
+
+const updateAttachment = (doc, updated, name, type) => {
+  const attachmentData = doc._attachments &&
+                         doc._attachments[name] &&
+                         doc._attachments[name].data &&
+                         doc._attachments[name].data.toString();
+  if (attachmentData === updated) {
+    return false;
+  }
+  doc._attachments[name] = {
+    data: Buffer.from(updated),
+    content_type: type
+  };
+  return true;
+};
+
+const updateAttachmentsIfRequired = (doc, updated) => {
+  const formUpdated = updateAttachment(doc, updated.form, 'form.html', 'text/html');
+  const modelUpdated = updateAttachment(doc, updated.model, 'model.xml', 'text/xml');
+  return formUpdated || modelUpdated;
+};
+
+const updateAttachments = (accumulator, doc) => {
+  return accumulator.then(results => {
+    const form = getEnketoForm(doc);
+    if (!form) {
+      results.push(null); // not an enketo form - no update required
+      return results;
+    }
+    logger.debug(`Generating html and xml model for enketo form "${doc._id}"`);
+    return module.exports._generate(form.data.toString()).then(result => {
+      results.push(result);
+      return results;
+    });
+  });
+};
+
+// Returns array of docs that need saving.
+const updateAllAttachments = docs => {
+  // spawn the child processes in series so we don't smash the server
+  return docs.reduce(updateAttachments, Promise.resolve([])).then(results => {
+    return docs.filter((doc, i) => {
+      const result = results[i];
+      if (!result) {
+        return false;
+      }
+      return updateAttachmentsIfRequired(doc, result);
```

Comment:
`result` is `undefined` in that suggestion. You can use `results[i]` instead, and because you picked up on this, I've gone for that solution. And yes, I'm totally happy with falsy! :)